src/data/list/basic.lean

```diff
@@ -2775,6 +2785,114 @@ theorem sublist.map (f : α → β) {l₁ l₂ : list α}
+  { simp only [reduce_option_nil, map_nil] },
+  { cases hd;
+    simpa only [true_and, option.map_some', map, eq_self_iff_true,
+                reduce_option_cons_of_some] using hl },
+end
+
+lemma reduce_option_append (l l' : list (option α)) :
+  (l ++ l').reduce_option = l.reduce_option ++ l'.reduce_option :=
+filter_map_append l l' id
+
+lemma reduce_option_length_le (l : list (option α)) :
+  l.reduce_option.length ≤ l.length :=
+begin
+  induction l with hd tl hl,
+  { simp only [reduce_option_nil, length] },
+  { cases hd,
+    { exact nat.le_succ_of_le hl },
+    { simpa only [length, add_le_add_iff_right, reduce_option_cons_of_some] using hl} }
+end
+
+lemma reduce_option_forall_iff {l : list (option α)} :
+  l.reduce_option.length = l.length ↔ ∀ x ∈ l, option.is_some x :=
+begin
+  induction l with hd tl hl,
+  { simp only [forall_const, reduce_option_nil, not_mem_nil,
+               forall_prop_of_false, eq_self_iff_true, length, not_false_iff] },
+  { cases hd,
+    { simp only [mem_cons_iff, forall_eq_or_imp, bool.coe_sort_ff, false_and,
+                 reduce_option_cons_of_none, length, option.is_some_none, iff_false],
+      intro H,
+      have := reduce_option_length_le tl,
+      rw H at this,
+      exact absurd (nat.lt_succ_self _) (not_lt_of_le this) },
+    { simp only [hl, true_and, mem_cons_iff, forall_eq_or_imp, add_left_inj,
+                 bool.coe_sort_tt, length, option.is_some_some, reduce_option_cons_of_some] } }
+end
+
+lemma reduce_option_exists_iff {l : list (option α)} :
+  l.reduce_option.length < l.length ↔ ∃ x ∈ l, x = none :=
+begin
+  induction l with hd tl hl,
+  { simp only [reduce_option_nil, not_mem_nil, length, not_false_iff, not_lt,
+               exists_false, exists_prop_of_false, iff_false] },
+  { cases hd,
+    { simp only [mem_cons_iff, reduce_option_cons_of_none, length, exists_prop,
+                 true_or, eq_self_iff_true, exists_eq_right, iff_true],
+      have := reduce_option_length_le tl,
+      exact lt_of_le_of_lt this (nat.lt_succ_self _) },
+    { simp only [hl, mem_cons_iff, length, reduce_option_cons_of_some,
+                 exists_prop, false_or, exists_eq_right, add_lt_add_iff_right] } }
+end
+
+lemma to_list_eq_reduce_option (x : option α) :
+  x.to_list = [x].reduce_option :=
+by { cases x; refl }
+
+lemma reduce_option_concat (l : list (option α)) (x : option α) :
+  (l.concat x).reduce_option = l.reduce_option ++ x.to_list :=
+begin
+  induction l with hd tl hl generalizing x,
+  { cases x;
+    simp [option.to_list] },
+  { simp only [concat_eq_append, reduce_option_append] at hl,
+    cases hd;
+    simp [hl, reduce_option_append] }
+end
+
+lemma reduce_option_concat_of_some (l : list (option α)) (x : α) :
+  (l.concat (some x)).reduce_option = l.reduce_option.concat x :=
+by simp only [reduce_option_nil, concat_eq_append, reduce_option_append, reduce_option_cons_of_some]
+
+lemma reduce_option_mem_iff {l : list (option α)} {x : α} :
```

Comment:
The name says to me that this iff should point the other direction, but I can't tell which order is better. My gut says to flip it...